README.md
# lx-helpers [![Build Status](https://img.shields.io/travis/litixsoft/lx-helpers/master.svg)](https://travis-ci.org/litixsoft/lx-helpers) [![david-dm](https://david-dm.org/litixsoft/lx-helpers.png)](https://david-dm.org/litixsoft/lx-helpers/) [![david-dm](https://david-dm.org/litixsoft/lx-helpers/dev-status.png)](https://david-dm.org/litixsoft/lx-helpers#info=devDependencies&view=table)

Litixsoft helper functions for Javascript (client and server).

## Getting Started

### Install:

[![NPM](https://nodei.co/npm/lx-helpers.png??downloads=true&stars=true)](https://nodei.co/npm/lx-helpers/)

## Documentation
[http://www.litixsoft.de/products-lxhelpers.html](http://www.litixsoft.de/products-lxhelpers.html) (german)

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Please create descriptive commit messages.
We use a git hook to validate the commit messages against these [rules](https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#heading=h.uyo6cb12dt6w).
Lint and test your code using [grunt](http://gruntjs.com/).

    $ grunt test

## Author
[Litixsoft GmbH](http://www.litixsoft.de)

## License
Copyright (C) 2013-2014 Litixsoft GmbH <info@litixsoft.de>
Licensed under the MIT license.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
<a name="0.5.1"></a>
### 0.5.1 (2014-03-19)


#### Features

* arrayIndexOf() and arrayHasItem can now handle prediacte functions as item ([269042da](https://github.com/litixsoft/lx-helpers/commit/269042da9bb0d48a0e57c01540edcf8183da1237))


<a name="0.5.0"></a>
## 0.5.0 (2014-02-19)


#### Features

* revert the params of the callback function in objectForEach() ([47199673](https://github.com/litixsoft/lx-helpers/commit/471996736b188b07c8ac57cd77b22dbc03838a6a))


#### Breaking Changes

* you have to change all calls to objectForEach()
 ([47199673](https://github.com/litixsoft/lx-helpers/commit/471996736b188b07c8ac57cd77b22dbc03838a6a))


<a name="v0.4.1"></a>
### v0.4.1 (2014-01-28)


#### Features

* add param index to callback function in forEach when iterating over arrays ([9a264e71](https://github.com/litixsoft/lx-helpers/commit/9a264e7186550b8291088c7908ba420dc563636a))

<a name="v0.4.0"></a>
## v0.4.0 (2014-01-28)


#### Features

* revert the params of the callback function in forEach ([fcf4c26d](https://github.com/litixsoft/lx-helpers/commit/fcf4c26d1e2b830b9b8f37c8024c34ff3ade3659))

#### Breaking Changes

* you have to change all calls to objectForEach and forEach which iterates over objects.
 ([fcf4c26d](https://github.com/litixsoft/lx-helpers/commit/fcf4c26d1e2b830b9b8f37c8024c34ff3ade3659))

<a name="v0.3.4"></a>
### v0.3.4 (2013-12-19)


#### Features

* add function removeTagsAndComments() ([44fdae46](https://github.com/litixsoft/lx-helpers/commit/44fdae4663e30a15fa03df8ff2a346454560bd23))
* add function isBoolean() ([998d6334](https://github.com/litixsoft/lx-helpers/commit/998d6334084dc482be8b433e8e0d89a782c2abf1))
* add function isRegExp() ([4de5b90b](https://github.com/litixsoft/lx-helpers/commit/4de5b90b8b340b15684c2b3c98b95db14a46702c))
* add function isString() ([f7af0beb](https://github.com/litixsoft/lx-helpers/commit/f7af0beb270a3590913dd0364c5b7f2cfa6376ff))

<a name="v0.3.3"></a>
### v0.3.3 (2013-12-18)


#### Bug Fixes

* rename module export in client side to window.lxHelpers ([6189db48](https://github.com/litixsoft/lx-helpers/commit/6189db481ab15ff9fb5be87f1f65a7090a342b30))

<a name="v0.3.2"></a>
### v0.3.2 (2013-12-09)


#### Features

* add function roundPrecise() to round  a value to n decimal places, with 0.005 be ([e9660ee6](https://github.com/litixsoft/lx-helpers/commit/e9660ee6502a61ca142646cfe6e9e3a364fb9107))

<a name="v0.3.1"></a>
### v0.3.1 (2013-12-09)


#### Features

* **Gruntfile:**
  * add grunt-bump task ([ec9e40a6](https://github.com/litixsoft/lx-helpers/commit/ec9e40a668c992bb5f5502ea482889290d1694e8))
  * add grunt-conventional-changelog to generate changelog ([cb73cc8d](https://github.com/litixsoft/lx-helpers/commit/cb73cc8d5c05eec39a0e8dada5b8a1ae27e34369))
* **git:** add git commit hook to validate the commit message ([778a6dce](https://github.com/litixsoft/lx-helpers/commit/778a6dce68298d4dfb30108b2aa1f5ffe1a7f30c))

## v0.3.0
#### Features
* add function isNumber()
* **getType():** now detects Infinite and NaN

#### Breaking Changes
* **getType():** now returns the type in lower case without the part ‘[object]‘

### v0.2.5
#### Features
* add break condition to forEach function

### v0.2.4
#### Features
* add function arrayHasItem(array, item) to check if an array contains the given item.

### v0.2.3
#### Features
* add function for getting a TypeError with a message.

### v0.2.2
#### Features
* function isEmpty() now only checks length for arrays and objects. Functions now return always true.

### v0.2.1
#### Features
* add function getType() to get the full type of the value (e.g. getType(null) === '[object Null]')

## v0.2.0
#### Features
* add zip file in dist folder containing the debug and min version of lx-helpers
* add function isEmpty() to check if the given value has no items/keys
* add function isArray() to check if the given value is an array
* add function isObject() to check if the given value is an object
* add function isFunction() to check if the given value is a function
* add function isDate() to check if the given value is a date
* add function forEach() to iterate over array, objects or functions
* update devDependencies

## v0.1.0 project initial

Gruntfile.js
'use strict';

var fs = require('fs');

module.exports = function (grunt) {
    // load npm tasks
    require('load-grunt-tasks')(grunt);

    // Project configuration.
    grunt.initConfig({
        jshintFiles: ['Gruntfile.js', 'lib/**/*.js', 'test/**/*.js'],
        pkg: grunt.file.readJSON('package.json'),
        banner: '/*!\n' +
            ' * <%= pkg.title || pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
            '<%= pkg.homepage ? " * " + pkg.homepage + "\\n" : "" %>' +
            ' *\n' +
            ' * Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>\n' +
            ' * Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %>\n' +
            ' */\n\n',
        // Before generating any new files, remove any previously-created files.
        clean: {
            reports: ['.reports'],
            coverage: ['.reports/coverage']
        },
        concat: {
            options: {
                stripBanners: true,
                banner: '<%= banner %>'
            },
            dist: {
                src: ['lib/<%= pkg.name %>.js'],
                dest: '<%= pkg.name %>.js'
            }
        },
        uglify: {
            options: {
                banner: '<%= banner %>'
            },
            dist: {
                src: ['<%= concat.dist.dest %>'],
                dest: '<%= pkg.name %>.min.js'
            }
        },
        jshint: {
            options: {
                jshintrc: true
            },
            test: '<%= jshintFiles %>',
            jslint: {
                options: {
                    reporter: 'jslint',
                    reporterOutput: '.reports/lint/jshint.xml'
                },
                files: {
                    src: '<%= jshintFiles %>'
                }
            },
            checkstyle: {
                options: {
                    reporter: 'checkstyle',
                    reporterOutput: '.reports/lint/jshint_checkstyle.xml'
                },
                files: {
                    src: '<%= jshintFiles %>'
                }
            }
        },
        bgShell: {
            coverage: {
                cmd: 'node node_modules/istanbul/lib/cli.js cover --dir .reports/coverage node_modules/grunt-jasmine-node/node_modules/jasmine-node/bin/jasmine-node -- test --forceexit'
            },
            cobertura: {
                cmd: 'node node_modules/istanbul/lib/cli.js report --root .reports/coverage --dir .reports/coverage/cobertura cobertura'
            }
        },
        open: {
            file: {
                path: '.reports/coverage/lcov-report/index.html'
            }
        },
        jasmine_node: {
            options: {
                specNameMatcher: './*.spec', // load only specs containing specNameMatcher
                requirejs: false,
                forceExit: true
            },
            test: ['test/'],
            ci: {
                options: {
                    jUnit: {
                        report: true,
                        savePath: '.reports/junit/',
                        useDotNotation: true,
                        consolidate: true
                    }
                },
                src: ['test/']
            }
        },
        changelog: {
            options: {
            }
        },
        bump: {
            options: {
                files: ['package.json', 'bower.json'],
                updateConfigs: ['pkg'],
                commitFiles: ['-a'],
                commitMessage: 'chore: release v%VERSION%',
                push: false
            }
        }
    });

    // Register tasks.
    grunt.registerTask('git:commitHook', 'Install git commit hook', function () {
        grunt.file.copy('validate-commit-msg.js', '.git/hooks/commit-msg');
        fs.chmodSync('.git/hooks/commit-msg', '0755');
        grunt.log.ok('Registered git hook: commit-msg');
    });

    grunt.registerTask('test', ['git:commitHook', 'jshint:test', 'jasmine_node:test']);
    grunt.registerTask('build', ['test', 'concat', 'uglify']);
    grunt.registerTask('cover', ['clean:coverage', 'jshint:test', 'bgShell:coverage', 'open']);
    grunt.registerTask('ci', ['clean', 'jshint:jslint', 'jshint:checkstyle', 'bgShell:coverage', 'bgShell:cobertura', 'jasmine_node:ci']);
    grunt.registerTask('release', 'Bump version, update changelog and tag version', function (version) {
        grunt.task.run([
            'bump:' + (version || 'patch') + ':bump-only',
            'build',
            'changelog',
            'bump-commit'
        ]);
    });

    // Default task.
    grunt.registerTask('default', 'build');
};

validate-commit-msg.js
#!/usr/bin node

/**
 * Git COMMIT-MSG hook for validating commit message
 * See https://docs.google.com/document/d/1rk04jEuGfk9kYzfqCuOlPTSJw3hEDZJTBN5E5f1SALo/edit
 *
 * Installation:
 * >> cd lx-helpers
 * >> ln -s validate-commit-msg.js .git/hooks/commit-msg
 */
var fs = require('fs');
var util = require('util');


var MAX_LENGTH = 100;
var PATTERN = /^(?:fixup!\s*)?(\w*)(\(([\w\$\.\-\*/]*)\))?\: (.*)$/;
var IGNORED = /^WIP\:/;
var TYPES = {
    feat: true,
    fix: true,
    docs: true,
    style: true,
    refactor: true,
    test: true,
    chore: true,
    revert: true
};


var error = function() {
    // gitx does not display it
    // http://gitx.lighthouseapp.com/projects/17830/tickets/294-feature-display-hook-error-message-when-hook-fails
    // https://groups.google.com/group/gitx/browse_thread/thread/a03bcab60844b812
    console.error('INVALID COMMIT MSG: ' + util.format.apply(null, arguments));
};


var validateMessage = function(message) {
    var isValid = true;

    if (IGNORED.test(message)) {
        console.log('Commit message validation ignored.');
        return true;
    }

    if (message.length > MAX_LENGTH) {
        error('is longer than %d characters !', MAX_LENGTH);
        isValid = false;
    }

    var match = PATTERN.exec(message);

    if (!match) {
        error('does not match "<type>(<scope>): <subject>" ! was: ' + message);
        return false;
    }

    var type = match[1];
    var scope = match[3];
    var subject = match[4];

    if (!TYPES.hasOwnProperty(type)) {
        error('"%s" is not allowed type !', type);
        return false;
    }

    // Some more ideas, do want anything like this ?
    // - allow only specific scopes (eg. fix(docs) should not be allowed ?
    // - auto correct the type to lower case ?
    // - auto correct first letter of the subject to lower case ?
    // - auto add empty line after subject ?
    // - auto remove empty () ?
    // - auto correct typos in type ?
    // - store incorrect messages, so that we can learn

    return isValid;
};


var firstLineFromBuffer = function(buffer) {
    return buffer.toString().split('\n').shift();
};



// publish for testing
exports.validateMessage = validateMessage;

// hacky start if not run by jasmine :-D
if (process.argv.join('').indexOf('jasmine-node') === -1) {
    var commitMsgFile = process.argv[2];
    var incorrectLogFile = commitMsgFile.replace('COMMIT_EDITMSG', 'logs/incorrect-commit-msgs');

    fs.readFile(commitMsgFile, function(err, buffer) {
        var msg = firstLineFromBuffer(buffer);

        if (!validateMessage(msg)) {
            fs.appendFile(incorrectLogFile, msg + '\n', function() {
                process.exit(1);
            });
        } else {
            process.exit(0);
        }
    });
}

lx-helpers.js
/*!
 * lx-helpers - v0.5.1 - 2014-03-19
 * https://github.com/litixsoft/lx-helpers
 *
 * Copyright (c) 2014 
 * Licensed MIT
 */

(function (exports) {
    'use strict';

    function isEmptyArray (array) {
        return array === undefined || array === null || (exports.isArray(array) && exports.isEmpty(array));
    }

    /**
     * Gets the type of the value in lower case.
     *
     * @param {*} value The value to test.
     * @returns {string}
     */
    exports.getType = function (value) {
        // inspired by http://techblog.badoo.com/blog/2013/11/01/type-checking-in-javascript/

        // handle null in old IE
        if (value === null) {
            return 'null';
        }

        // handle DOM elements
        if (value && (value.nodeType === 1 || value.nodeType === 9)) {
            return 'element';
        }

        var s = Object.prototype.toString.call(value);
        var type = s.match(/\[object (.*?)\]/)[1].toLowerCase();

        // handle NaN and Infinity
        if (type === 'number') {
            if (isNaN(value)) {
                return 'nan';
            }
            if (!isFinite(value)) {
                return 'infinity';
            }
        }

        return type;
    };

    /**
     * Returns if the value is an array.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isArray = function (value) {
        return exports.getType(value) === 'array';
    };

    /**
     * Returns if the value is an object.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isObject = function (value) {
        return exports.getType(value) === 'object';
    };

    /**
     * Returns if the value is a function.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isFunction = function (value) {
        return exports.getType(value) === 'function';
    };

    /**
     * Returns if the value is a Date.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isDate = function (value) {
        return exports.getType(value) === 'date';
    };

    /**
     * Returns if the value is a Number.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isNumber = function (value) {
        return exports.getType(value) === 'number';
    };

    /**
     * Returns if the value is a string.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isString = function (value) {
        return exports.getType(value) === 'string';
    };

    /**
     * Returns if the value is a regular expression object.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isRegExp = function (value) {
        return exports.getType(value) === 'regexp';
    };

    /**
     * Returns if the value is a boolean.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isBoolean = function (value) {
        return exports.getType(value) === 'boolean';
    };

    /**
     * Returns if the value has no items/keys.
     *
     * @param {*} value The value to test.
     * @returns {boolean}
     */
    exports.isEmpty = function (value) {
        if (exports.isArray(value)) {
            return value.length === 0;
        }

        if (exports.isObject(value)) {
            return Object.keys(value).length === 0;
        }

        return true;
    };

    /**
     * Returns the first item of the array which matchs the predicate.
     *
     * @param {Array} array The array.
     * @param {Function} predicate The predicate function.
     * @param {Object=} context The context of this.
     * @return {*}
     */
    exports.arrayFirst = function (array, predicate, context) {
        if (isEmptyArray(array)) {
            return null;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        for (var i = 0, j = array.length; i < j; i++) {
            if (predicate.call(context, array[i])) {
                return array[i];
            }
        }

        return null;
    };

    /**
     * Perform action for each array item.
     *
     * @param {Array} array The array.
     * @param {function(item, index)} action The action to perform with each array item.
     * @param {Object=} context The context of this.
     */
    exports.arrayForEach = function (array, action, context) {
        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        return exports.forEach(array, action, context);
    };

    /**
     * Returns the index of the item in the array.
     *
     * @param {Array} array The array.
     * @param {*} item The item to search for.
     * @param {Object=} context The context of this.
     * @return {Number}
     */
    exports.arrayIndexOf = function (array, item, context) {
        if (isEmptyArray(array)) {
            return -1;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        // handle predicate function
        if (exports.isFunction(item)) {
            for (var a = 0, b = array.length; a < b; a++) {
                if (item.call(context, array[a])) {
                    return a;
                }
            }
        } else {
            if (typeof Array.prototype.indexOf === 'function') {
                return Array.prototype.indexOf.call(array, item);
            }

            for (var i = 0, j = array.length; i < j; i++) {
                if (array[i] === item) {
                    return i;
                }
            }
        }

        return -1;
    };

    /**
     * Returns if the item is in the array.
     *
     * @param {Array} array The array.
     * @param {*} item The item to search for.
     * @param {Object=} context The context of this.
     * @return {Boolean}
     */
    exports.arrayHasItem = function (array, item, context) {
        return exports.arrayIndexOf(array, item, context) > -1;
    };

    /**
     * Removes the item from the array.
     *
     * @param {Array} array The array.
     * @param {*} itemToRemove The item to remove.
     * @param {Object=} context The context of this.
     */
    exports.arrayRemoveItem = function (array, itemToRemove, context) {
        if (isEmptyArray(array)) {
            return;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        var index = exports.arrayIndexOf(array, itemToRemove, context);

        if (index >= 0) {
            array.splice(index, 1);
        }
    };

    /**
     * Applies the mapping function on each item of the array and return the mapped array.
     *
     * @param {Array} array The array.
     * @param {Function} mapping The mapping function.
     * @param {Object=} context The context of this.
     * @return {Array}
     */
    exports.arrayMap = function (array, mapping, context) {
        var result = [];

        if (isEmptyArray(array)) {
            return result;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        for (var i = 0, j = array.length; i < j; i++) {
            result.push(mapping.call(context, array[i]));
        }

        return result;
    };

    /**
     * Gets the distinct values of an array.
     *
     * @param {Array} array The array.
     * @param {Object=} context The context of this.
     * @return {Array}
     */
    exports.arrayGetDistinctValues = function (array, context) {
        var result = [];

        if (isEmptyArray(array)) {
            return result;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        for (var i = 0, j = array.length; i < j; i++) {
            if (exports.arrayIndexOf(result, array[i], context) < 0) {
                result.push(array[i]);
            }
        }

        return result;
    };

    /**
     * Filters an array.
     *
     * @param {Array} array The array.
     * @param {Function} predicate The predicate function.
     * @param {Object=} context The context of this.
     * @return {Array}
     */
    exports.arrayFilter = function (array, predicate, context) {
        var result = [];

        if (isEmptyArray(array)) {
            return result;
        }

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        for (var i = 0, j = array.length; i < j; i++) {
            if (predicate.call(context, array[i])) {
                result.push(array[i]);
            }
        }

        return result;
    };

    /**
     * Pushes the values to the array in one step.
     *
     * @param {Array} array The array.
     * @param {Array} valuesToPush The values to push in the array.
     * @param {Boolean=} clearArray Should the array be cleared before pushing the new values.
     * @return {Array}
     */
    exports.arrayPushAll = function (array, valuesToPush, clearArray) {
        array = array || [];

        if (!exports.isArray(array)) {
            throw exports.getTypeError('array', array, []);
        }

        if (!exports.isArray(valuesToPush)) {
            throw exports.getTypeError('valuesToPush', valuesToPush, []);
        }

        if (clearArray) {
            array.length = 0;
        }

        array.push.apply(array, valuesToPush);

        return array;
    };

    /**
     * Performs action for each item in object.
     *
     * @param {Object} obj The object.
     * @param {function(key, value)} action The action to perform with each object item.
     * @param {Object=} context The context of this.
     */
    exports.objectForEach = function (obj, action, context) {
        if (!exports.isObject(obj)) {
            throw exports.getTypeError('obj', obj, {});
        }

        if (!exports.isFunction(action)) {
            throw exports.getTypeError('action', action, function () {});
        }

        var i;
        var keys = Object.keys(obj);
        var length = keys.length;

        for (i = 0; i < length; i++) {
            if (action.call(context, keys[i], obj[keys[i]]) === false) {
                break;
            }
        }
    };

    /**
     * Performs action for each item.
     *
     * @param {Array|Object|Function} obj The object with items.
     * @param {function(value, key)} action The action to perform with each item.
     * @param {Object=} context The context of this.
     */
    exports.forEach = function (obj, action, context) {
        if (exports.isEmpty(obj)) {
            return;
        }

        if (!exports.isFunction(action)) {
            throw exports.getTypeError('action', action, function () {});
        }

        var i, length, keys;

        if (exports.isArray(obj)) {
            for (i = 0, length = obj.length; i < length; i++) {
                if (action.call(context, obj[i], i) === false) {
                    break;
                }
            }
        } else if (exports.isObject(obj)) {
            keys = Object.keys(obj);
            length = keys.length;

            for (i = 0; i < length; i++) {
                if (action.call(context, obj[keys[i]], keys[i]) === false) {
                    break;
                }
            }
        }
    };

    /**
     * Creates a new GUID
     *
     * @returns {string}
     */
    exports.createGuid = function () {
        var S4 = function () {
            return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
        };

        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
    };

    /**
     * Capitalize the first letter of a word and returns the word.
     *
     * @param {String} word The word.
     * @return {String} The capitalized word.
     */
    exports.capitalizeFirstLetter = function (word) {
        word = word || '';

        if (typeof word !== 'string') {
            throw exports.getTypeError('word', word, '');
        }

        return word.charAt(0).toUpperCase() + word.slice(1);
    };

    /**
     * Clones a javascript object, array or primitive
     *
     * @param {*} src The source.
     * @returns {*}
     */
    exports.clone = function (src) {
        var result = {};

        function mixin (dest, source, cloneFunc) {
            if (exports.isObject(source)) {
                exports.forEach(source, function (value, key) {
                    dest[key] = cloneFunc ? cloneFunc(value) : value;
                });
            }

            return dest;
        }

        if (!src || typeof src !== 'object' || typeof src === 'function') {
            return src;
        }

        // Date
        if (src instanceof Date) {
            return new Date(src.getTime());
        }

        // RegExp
        if (src instanceof RegExp) {
            return new RegExp(src);
        }

        // Array
        if (exports.isArray(src)) {
            result = [];

            exports.arrayForEach(src, function (item) {
                result.push(exports.clone(item));
            });
        }

        return mixin(result, src, exports.clone);
    };

    /**
     * Gets a TypeError with a message.
     *
     * @param {!String} name The name of the value.
     * @param {!*} value The value.
     * @param {!*} expectedType The type which is expected
     * @returns {TypeError}
     */
    exports.getTypeError = function (name, value, expectedType) {
        return new TypeError('Param "' + name + '" is of type ' + exports.getType(value) + '! Type ' + exports.getType(expectedType) + ' expected');
    };

    /**
     * Returns the value of a rounded to two decimal places, with 0.005 being rounded up.
     *
     * @param {number} value The value.
     * @param {number} decimals The decimal places
     * @returns {number | NaN}
     */
    exports.roundPrecise = function (value, decimals) {
        // if value is not a number or the decPlaces is not an integer.
        if ((value === null || isNaN(value)) || !(typeof decimals === 'number' && decimals % 1 === 0)) {
            return NaN;
        }

        var val = value * Math.pow(10, decimals);
        var fraction = (Math.round((val - parseInt(val)) * 10) / 10);

        // this line is for consistency with .NET Decimal.Round behavior => -342.055 => -342.06
        if (fraction === -0.5) {
            fraction = -0.6;
        }

        val = Math.round(parseInt(val) + fraction) / Math.pow(10, decimals);
        return val;
    };

    /**
     * Removes all tags and comments from a html string.
     *
     * @param {!string} input The input string.
     * @param {string=} allowedTags The srring with allowed tags which will not be removed.
     * @returns {string}
     */
    exports.removeTagsAndComments = function (input, allowedTags) {
        allowedTags = allowedTags || '';

        if (!exports.isString(input)) {
            throw exports.getTypeError('input', input, '');
        }

        if (!exports.isString(allowedTags)) {
            throw exports.getTypeError('allowed', allowedTags, '');
        }

        // making sure the allowed arg is a string containing only tags in lowercase (<a><b><c>)
        allowedTags = (allowedTags.toLowerCase().match(/<[a-z][a-z0-9]*>/g) || []).join('');

        var tags = /<\/?([a-z][a-z0-9]*)\b[^>]*>/gi;
        var comments = /<!--[\s\S]*?-->/gi;

        return input.replace(comments, '').replace(tags, function ($0, $1) {
            return allowedTags.indexOf('<' + $1.toLowerCase() + '>') > -1 ? $0 : '';
        });
    };

})(typeof(module) !== 'undefined' && module.exports !== undefined ? module.exports : (window.lxHelpers = window.lxHelpers || {}));